feat(serenity-bdd): include Serenity BDD CLI jar file into files of the published package

fapspirit opened the pull request:
This change should simplify overall setup experience for most of the cases by eleminating the need to run "serenity-bdd update" manually. Default version of Serenity BDD will be saved into default directory and will be included into files under the default path of the package and published to NPM registry.

This change does not affect current customization capabilities in any way, it's still possible to set custom cache directory or set specific version of Serenity BDD.

Related tickets: #2560

fapspirit commented:
Hey @jan-molak, sorry for the delay with the implementation, finally found the time to finish what I started.

This implementation mostly follows our plan that we agreed on earlier during the discussion in the ticket. The one difference is when `serenity-bdd update` command is called: we've agreed on to call it in `postinstall` hook, however it's not possible on a clean install due to code is not yet compiled from ts to js at this stage, so I've decided to call it as part of the `compile` command. Please take a look and let me know what you think.

jan-molak reviewed (commented):
Hi @fapspirit, I'm sorry for not getting back to you sooner. I was on holiday with limited access to the Internet.

The PR looks good to me. I have a few suggestions, so I'd love to hear your thoughts.

There are also some updates to Serenity/JS documentation and examples we'll need to take care of when merging your PR:
- [ ] update `serenity-js.org` installation instructions to remove the [recommendation to use `serenity-bdd update`](https://github.com/search?q=repo%3Aserenity-js%2Fserenity-js.org+content%3A%22serenity-bdd+update%22&type=code)
- [ ] update Serenity/JS project templates to remove `postinstall` scripts, there are [11 of them](https://github.com/search?q=org%3Aserenity-js+AND+path%3A%22package.json%22+AND+content%3A%2F%22name%22%3A+%22.*-template%22%2F+AND+content%3A%22serenity-bdd+update%22+&type=code)
- [ ] update the [Serenity/JS Serenity BDD Reporter docs](https://github.com/serenity-js/serenity-js.org/blob/main/src/docs/handbook/reporting/serenity-bdd-reporter.mdx) to explain that running the `serenity-bdd update` command is optional, but can be kept in existing implementations if needd 

Would you be interested in tackling any of the above?

jan-molak commented on packages/serenity-bdd/package.json:
```diff
@@ -38,7 +38,8 @@
     "test:no-coverage": "mocha --config ../../.mocharc.yml 'spec/**/*.spec.*'",
-    "compile": "rimraf lib && tsc --project tsconfig.build.json"
+    "compile": "rimraf lib cache && tsc --project tsconfig.build.json && npm run populate-cache",
+    "populate-cache": "./bin/serenity-bdd update --cacheDir './cache'"
```
By the way, if you'd like to commit these changes, please feel free to mark them as `fix` or `chore`; your initial `feat` commit will "win" when the CI pipeline determines the build number and bump the `minor` version.

fapspirit replied:
> If you renamed populate-cache to postcompile, NPM would automatically call this [post-script](https://docs.npmjs.com/cli/v9/using-npm/scripts#pre--post-scripts) when compile passes.

Hmm, this actually looks like a more ergonomic way, I'll try it and probably will stick with the `post` hook so we can avoid introduction of the additional explicit step in the CI.
By the way, `pre` and `post` hooks are stable in the npm, however they're not supported by [yarn](https://yarnpkg.com/advanced/lifecycle-scripts) and was disabled by default in [pnpm](https://github.com/pnpm/pnpm/issues/2891) (although reverted back in [pnpm v9](https://github.com/pnpm/pnpm/pull/7634)). I think it's not a problem for SerenityJS as it's using npm, but might be useful if you'll ever consider to switch the package manager for the project.

> I'd keep the --ignoreSSL flag for contributors whose network setups struggle with correct handling of SSL certificates. Unfortunatelly, there's still plenty of broken corporate networks out there.

Nice catch, didn't considered this case. Will add this flag.

> I'd add node to make the interpreter explicit.

I've decided to not use `node` as it was my assumption that it works that way when executed via CLI by the user. In that case, it looks like an implicit integration test. But I don't insist on that as I'm actually not sure in the correctness of my assumption and I can change this script to call `node` explicitly to make it more reliable.

jan-molak replied:
> By the way, pre and post hooks are stable in the npm, however they're not supported by yarn and was disabled by default in pnpm (although reverted back in pnpm v9). I think it's not a problem for SerenityJS as it's using npm, but might be useful if you'll ever consider to switch the package manager for the project.

Oh, that's a good tip, thank you. I wanted to investigate using pnpm instead of the regular npm since new versions of nx (replacement of lerna) don't support hoisting or workspaces. I'll keep your advice in mind.

> I've decided to not use node as it was my assumption that it works that way when executed via CLI by the user.

Windows-based machines might sometimes get confused, adding an explicit call to `node` helps to avoid that.

fapspirit commented:
> There are also some updates to Serenity/JS documentation and examples we'll need to take care of when merging your PR:
[...]
Would you be interested in tackling any of the above?

Sure, I'll work on these changes. I'm not that good at tech writing though, I'll appreciate some help with the docs. Perhaps you can own the docs part and I'll take the rest?

fapspirit commented on packages/serenity-bdd/package.json:
```diff
@@ -38,7 +38,8 @@
     "test": "c8 npm run test:no-coverage",
     "test:no-coverage": "mocha --config ../../.mocharc.yml 'spec/**/*.spec.*'",
-    "compile": "rimraf lib && tsc --project tsconfig.build.json"
+    "compile": "rimraf lib cache && tsc --project tsconfig.build.json && npm run populate-cache",
```
> I'd suggest to use rimraf lib instead of rimraf lib cache and keep the local cache since the jars are immutable and don't need to be redownloaded

Good point, makes sense to me, will adjust.

> Populating the cache feels like a separate step to compile, I'd suggest removing the explicit call to `npm run populate-cache`

I'm mot entirely sure I get the idea, please correct me if I'm wrong. You propose to explicitly add `populate-cache` step in the build process in the [Makefile](https://github.com/serenity-js/serenity-js/blob/cc6e399a6ec5433e7470eb8bd0267479f50c4989/Makefile#L40) and in the [CI](https://github.com/serenity-js/serenity-js/blob/cc6e399a6ec5433e7470eb8bd0267479f50c4989/.github/workflows/main.yaml#L29)?

jan-molak commented:
> > There are also some updates to Serenity/JS documentation and examples we'll need to take care of when merging your PR:
> > [...]
> > Would you be interested in tackling any of the above?
> 
> Sure, I'll work on these changes.

Fab, thank you!

> I'm not that good at tech writing though, I'll appreciate some help with the docs. Perhaps you can own the docs part and I'll take the rest?

Of course 👍🏻

fapspirit commented:
Hey @jan-molak, I've implemented changes as you suggested:
1. Moved `npm run populate-cache` from `compile` script to `postcompile` hook
2. `cache` directory is persisted across invocations of the `compile` script
3. Added explicit call of `node` in the `populate-cache` script

Also I've removed `postinstall` script from all of the templates:
- https://github.com/serenity-js/serenity-js-cucumber-webdriverio-template/pull/767
- https://github.com/serenity-js/serenity-js-playwright-test-template/pull/428
- https://github.com/serenity-js/serenity-js-cucumber-template/pull/795
- https://github.com/serenity-js/serenity-js-cucumber-protractor-template/pull/1295
- https://github.com/serenity-js/serenity-js-playwright-ct-react-template/pull/197
- https://github.com/serenity-js/serenity-js-cucumber-playwright-template/pull/429
- https://github.com/serenity-js/serenity-js-mocha-webdriverio-template/pull/918
- https://github.com/serenity-js/serenity-js-playwright-ct-web-components-template/pull/37
- https://github.com/serenity-js/serenity-js-mocha-template/pull/825
- https://github.com/serenity-js/serenity-js-jasmine-webdriverio-template/pull/281
- https://github.com/serenity-js/protractor-cucumber-framework-demo/pull/187

And for the last part, I've removed `postinstall` hook from the docs: https://github.com/serenity-js/serenity-js.org/pull/43
I've removed all related steps from examples, but left the section that describes `update` command. I think this section might need some additional docs or examples of cases when the `update` command might be useful because now it's not quite clear why this command ever needed since it's no longer used in the handbook.
